fix(form): improve reference error states

bjoerge opened the pull request:
### Description

If you have a reference to a nonexistent document, we currently display a confusing/unhelpful message. This PR refactors some of the logic we already had for handling the various cases that may occur with references to make them show up as originally intended.

**Before**
<img width="641" alt="image" src="https://github.com/sanity-io/sanity/assets/876086/7aec06e0-7407-4637-86a4-cbca68b00d4e">

**After**
<img width="682" alt="image" src="https://github.com/sanity-io/sanity/assets/876086/dacf937c-fc46-4432-8707-763a2d84531e">

(Note: the reference strength mismatch is irrelevant in this case, since it's been deleted while still being referenced it can only be a weak ref, and can't be made into a strong reference)


Also snuck in a fix for an issue causing references of invalid types to never load:
**Before**
<img width="659" alt="image" src="https://github.com/sanity-io/sanity/assets/876086/f9d1b1e0-a000-4feb-93ef-df6e09994ab3">
(forever)

**After**
<img width="795" alt="image" src="https://github.com/sanity-io/sanity/assets/876086/43b48f4e-251a-4a36-b786-d4bd4d010908">
(note: "empty" is the actual type name here)


### What to review

Make sure error states for reference inputs renders correctly:
- A (weak) reference that can't be found
- Referenced document is inaccessible by current role (this one might be hard to test, so feel free to skip)
- Weak/strong mismatch between schema and value
- Referenced document is of invalid type

### Notes for release

- Fixed an issue where references to undeclared document types would never load
- Improved error UIs for reference values

rexxars commented on packages/sanity/src/core/form/inputs/ReferenceInput/PreviewReferenceValue.tsx:
```diff
@@ -81,3 +153,45 @@ export function PreviewReferenceValue(props: {
+  return [
+    ...subList.map((item, i) => <Fragment key={i}>{item}, </Fragment>),
+    <Fragment key="last">
+      {conjunction} {list.at(-1)}
```
We still use ES2017 as our build target, but `Array.prototype.at` is not included until ES2022.
```suggestion
      {conjunction} {list[list.length - 1]}
```

robinpyon commented on packages/sanity/src/core/form/inputs/ReferenceInput/PreviewReferenceValue.tsx:
```diff
@@ -57,21 +59,91 @@ export function PreviewReferenceValue(props: {
-        The referenced document is of invalid type: ({refTypeName || 'unknown'})
-        <pre>{JSON.stringify(value, null, 2)}</pre>
-      </Stack>
+      <Inline space={2}>
+        <Box padding={1}>
+          <Flex align="center">
+            <Box flex={1} paddingY={2}>
+              <Text muted>Type of referenced document not declared in schema</Text>
+            </Box>
+          </Flex>
+        </Box>
+        <Box>
+          <Tooltip
+            portal
+            content={
+              <Stack space={3} padding={3}>
+                <Text size={1}>
+                  Referenced document (<code>{value._ref}</code>) is of type{' '}
+                  <code>{referenceInfo.result.type}</code>.
+                </Text>
+                <Text size={1}>
+                  According to the schema, referenced documents can only be of type{' '}
+                  {humanizeList(
+                    type.to.map((toType) => <code key={toType.name}>{toType.name}</code>),
+                    'or'
+                  )}
+                </Text>
+              </Stack>
+            }
+          >
+            <TextWithTone tone="default">
+              <HelpCircleIcon />
+            </TextWithTone>
+          </Tooltip>
+        </Box>
+      </Inline>
```
```suggestion
      <Flex align="center" justify="flex-start">
        <Box padding={1}>
          <Flex align="center">
            <Box flex={1} paddingY={2}>
              <Text muted>Type of referenced document not declared in schema</Text>
            </Box>
          </Flex>
        </Box>
        <Box>
          <Tooltip
            portal
            content={
              <Stack space={3} padding={3}>
                <Text size={1}>
                  Referenced document (<code>{value._ref}</code>) is of type{' '}
                  <code>{referenceInfo.result.type}</code>.
                </Text>
                <Text size={1}>
                  According to the schema, referenced documents can only be of type{' '}
                  {humanizeList(
                    type.to.map((toType) => <code key={toType.name}>{toType.name}</code>),
                    'or'
                  )}
                  .
                </Text>
              </Stack>
            }
          >
            <Box padding={2}>
              <TextWithTone tone="default">
                <HelpCircleIcon />
              </TextWithTone>
            </Box>
          </Tooltip>
        </Box>
      </Flex>
```

**suggestion**: Using a wrapping `<Flex>` rather than `<Inline>` to ensure that the icon / context menu wraps gracefully at smaller breakpoints, e.g.

<img width="379" alt="image" src="https://github.com/sanity-io/sanity/assets/209129/fc82605f-5f09-437c-9c98-dec39de4cc15">

vs.

<img width="381" alt="image" src="https://github.com/sanity-io/sanity/assets/209129/8c6bbde0-7d18-4299-bcb0-708569be461f">

mikolajdobrucki replied:
@robinpyon which one is 'before' and which one is 'after' your change?
I guess that the icon should *not* wrap under the text?

robinpyon replied:
That's correct – first image is 'after', this could have been clearer!

robinpyon commented on packages/sanity/src/core/form/inputs/ReferenceInput/PreviewReferenceValue.tsx:
```diff
@@ -57,21 +59,91 @@ export function PreviewReferenceValue(props: {
+                    (id: <code>{value._ref}</code>)
+                  </UnavailableMessage>
+                ) : (
+                  <UnavailableMessage title="Insufficcient permissions" icon={AccessDeniedIcon}>
```
```suggestion
                  <UnavailableMessage title="Insufficient permissions" icon={AccessDeniedIcon}>
```

robinpyon reviewed (approved):
Thanks @bjoerge! Left a few comments, nothing major, all cosmetic

I did also notice that references that both have a reference strength AND schema type mismatch will not render a title, e.g. (note `""-document` )

<img width="352" alt="image" src="https://github.com/sanity-io/sanity/assets/209129/d17fe29a-2fb1-43e2-9d83-957e29ff25a9">

Ideally this would fall back to a generic title (e.g. "this document"), but this is super edge case stuff

rexxars commented on packages/sanity/src/core/form/inputs/ReferenceInput/PreviewReferenceValue.tsx:
```diff
@@ -81,3 +153,45 @@ export function PreviewReferenceValue(props: {
+  )
+}
+
+const humanizeList = (list: React.ReactNode[], conjunction: string) => {
```
> **question:** what qualifies for inclusion in `@sanity/utils`?

Pretty sure we want to move away from `@sanity/utils` altogether. Utilities should either be internal to the `sanity` module, or exported publicly from the `sanity` module if they're reusable enough that it makes sense for our users to use them.

bjoerge replied:
It just didn't feel worth it to include an external dependency for just this single case. Should be easy to replace with something from npm or turn it into it's own package at a later point if we need it elsewhere too.

mikolajdobrucki commented on packages/sanity/src/core/form/inputs/ReferenceInput/PreviewReferenceValue.tsx:
```diff
@@ -57,21 +59,94 @@ export function PreviewReferenceValue(props: {
+        <Box padding={1}>
+          <Flex align="center">
+            <Box flex={1} paddingY={2}>
+              <Text muted>Type of referenced document not declared in schema</Text>
```
I would suggest to change this copy to "Document of invalid type"

bjoerge replied:
Agreed - fixed by 544c368153

mikolajdobrucki reviewed (approved):
Amazing work! Looks good to me! 🚀

bjoerge commented:
Thanks for review everyone :)


@robinpyon - and thanks for catching the ""-document issue - I've fixed it so it display the document title if available, or "the referenced document" if not:
<img width="598" alt="image" src="https://github.com/sanity-io/sanity/assets/876086/ca28c4a8-c4b7-4689-be14-0c63ab8bae71">

<img width="591" alt="image" src="https://github.com/sanity-io/sanity/assets/876086/e8f0acef-8773-4f6f-8b4a-33fc5ddbbc8a">

While doing a last round of testing I found another visual issue (also very edge casey, but still):
If you create a new reference in place of a certain document type and later remove that type from the list of valid reference types in the schema, it would show up like this:
![image](https://github.com/sanity-io/sanity/assets/876086/69040d89-2644-45f5-aa51-e4bac999fa52)

I've fixed this in 544897bbff so it looks and behaves the same as the other type mismatch message:
![image](https://github.com/sanity-io/sanity/assets/876086/e1195eb0-031e-489a-bfbd-3017db962392)
